### ISSUE TITLE
Phase 3: Answer Verifier (AST-Aware QA Model, 15M params)

### ISSUE THREAD

**Issue by marctjones** [marctjones/klareco#583]
# Phase 3: Answer Verifier

**Parent Epic**: #579  
**Depends on**: #582 (Phase 2 role labeler)

## Goal

Train a ~15M param model to verify if a candidate answer actually answers the question in context. This is the final semantic validation layer that catches wrong answers even when entity type and role are correct.

## Problem Examples

**Current behavior:**
```
Query: "Kiu kreis Esperanton?"
Doc #1: "esperantistoj en Berlino"
Candidate: "Esperantista"
  - Entity type: PERSON_GENERIC ✓ (correct type)
  - Semantic role: Could be AGENT ✓
  - Problem: Context is about "Esperantists in Berlin", NOT about creation!

Should verify: Does "Esperantista" answer "Who created Esperanto?" → NO (0.05 relevance)
```

**Another example:**
```
Query: "Kiam estis kreita Esperanto?"
Doc #3: "La Retbutiko de FEL estis kreita en 1996"
Candidate: "1996"
  - Entity type: TIME_ABSOLUTE ✓
  - Semantic role: TIME ✓
  - Problem: Context is about FEL webshop, not Esperanto creation!

Should verify: Does "1996" answer "When was Esperanto created?" → NO (0.10 relevance)
Correct answer: "1887" (different document)
```

### Input (AST-Based Cross-Encoder)

```python
{
    # Question representation
    "question": {
        "text": "Kiu kreis Esperanton?",
        "ast": {
            "tipo": "frazo",
            "fraztipo": "demando",
            "verbo": {"radiko": "kre"},
            "objekto": {"kerno": {"radiko": "Esperant"}}
        },
        "question_type": "WHO",
        "target_entity": "Esperant",  # What the question is about
        "target_role": "PATIENT"      # What role it plays
    },
    
    # Candidate answer
    "answer": {
        "text": "Esperantista",
        "ast": {
            "tipo": "vorto",
            "radiko": "esperant",
            "sufiksoj": ["ist"],
            "vortspeco": "substantivo"
        },
        "entity_type": "PERSON_GENERIC",  # From Phase 1
        "semantic_role": "AGENT"           # From Phase 2
    },
    
    # Context (document snippet)
    "context": {
        "text": "esperantistoj en Berlino",
        "ast": {
            "tipo": "frazo",
            "subjekto": {"kerno": {"radiko": "esperantist"}},
            "aliaj": [
                {"radiko": "en"},
                {"radiko": "Berlin"}
            ]
        },
        "snippet_position": "sentence_3_of_10"
    }
}
```

### Output

```python
{
    "relevance_score": 0.05,  # [0-1] Does answer match question in context?
    "confidence": 0.93,         # Model confidence
    "explanation": "Context discusses Esperantists in Berlin (location/group), not creation event. Verb mismatch: no creation verb in context.",
    
    # Breakdown (for debugging)
    "scores": {
        "semantic_match": 0.02,      # Does answer semantically fit question?
        "context_relevance": 0.05,   # Is context about the right topic?
        "verb_alignment": 0.08,      # Do verbs match?
        "entity_consistency": 0.95   # Is entity the same? (Esperant- stem)
    },
    
    # Features
    "features": {
        "question_verb": "kre",
        "context_verb": null,         # No verb in snippet
        "question_entity": "Esperant",
        "context_entities": ["esperantist", "Berlin"],
        "entity_overlap": True,       # "Esperant-" stem overlap
        "verb_match": False,
        "context_type": "location_description"  # Not creation event
    }
}
```

### Model Design (Cross-Encoder + AST Features)

```python
class AnswerVerifier(nn.Module):
    """
    Cross-encoder that jointly encodes question, answer, and context.
    Uses both text and AST representations.
    """
    
    def __init__(self):
        # Question encoder
        self.question_encoder = QuestionEncoder(
            ast_encoder_dim=128,
            question_type_dim=16
        )
        
        # Answer encoder
        self.answer_encoder = AnswerEncoder(
            ast_encoder_dim=128,
            entity_type_dim=16,
            role_dim=16
        )
        
        # Context encoder
        self.context_encoder = ContextEncoder(
            ast_encoder_dim=128,
            max_context_len=100
        )
        
        # Cross-attention layers (question-answer, question-context, answer-context)
        self.qa_attention = CrossAttention(128, num_heads=4)
        self.qc_attention = CrossAttention(128, num_heads=4)
        self.ac_attention = CrossAttention(128, num_heads=4)
        
        # Feature extractor (deterministic features)
        self.feature_extractor = VerificationFeatureExtractor()
        
        # Verification head
        self.verifier = nn.Sequential(
            nn.Linear(128*3 + 64, 256),  # Q+A+C + features
            nn.ReLU(),
            nn.Dropout(0.3),
            nn.Linear(256, 128),
            nn.ReLU(),
            nn.Linear(128, 1),  # Single relevance score
            nn.Sigmoid()
        )
    
    def forward(self, question, answer, context):
        # Encode each component
        q_emb = self.question_encoder(question['ast'], question['question_type'])
        a_emb = self.answer_encoder(
            answer['ast'], 
            answer['entity_type'], 
            answer['semantic_role']
        )
        c_emb = self.context_encoder(context['ast'])
        
        # Cross-attention (learn relationships)
        qa_attn = self.qa_attention(q_emb, a_emb)
        qc_attn = self.qc_attention(q_emb, c_emb)
        ac_attn = self.ac_attention(a_emb, c_emb)
        
        # Extract deterministic features
        features = self.feature_extractor(question, answer, context)
        
        # Combine and verify
        combined = torch.cat([qa_attn, qc_attn, ac_attn, features])
        relevance = self.verifier(combined)
        
        return relevance
```

### Question Encoder

```python
class QuestionEncoder(nn.Module):
    """Encode question AST + question type."""
    
    def __init__(self, ast_encoder_dim=128, question_type_dim=16):
        super().__init__()
        
        # AST encoder (reuse from previous phases)
        self.ast_encoder = ASTGraphEncoder(node_dim=ast_encoder_dim)
        
        # Question type embedding
        self.question_types = ['WHO', 'WHAT', 'WHERE', 'WHEN', 'HOW_MANY', 'WHY', 'HOW']
        self.question_type_embedding = nn.Embedding(len(self.question_types), question_type_dim)
        
        # Combine
        self.combiner = nn.Linear(ast_encoder_dim + question_type_dim, ast_encoder_dim)
    
    def forward(self, question_ast, question_type):
        # Encode AST
        ast_emb = self.ast_encoder(question_ast)
        
        # Encode question type
        type_idx = self.question_types.index(question_type)
        type_emb = self.question_type_embedding(torch.tensor(type_idx))
        
        # Combine
        combined = torch.cat([ast_emb, type_emb])
        return self.combiner(combined)
```

### Data Generation Strategy

```python
# 1. Positive examples (5000)
# - Correct (question, answer, context) triples from corpus
# - Label: 1.0 (high relevance)

# 2. Hard negatives (5000)
# - Wrong answers from SAME document (same topic, wrong answer)
# - Example: Q="Kiu kreis?" A="Esperantista" Context="esperantistoj..."
# - Label: 0.0-0.3 (low relevance)

# 3. Easy negatives (3000)
# - Random answers from different documents (different topic)
# - Example: Q="Kiu kreis?" A="Germanio" Context="geography text"
# - Label: 0.0 (no relevance)

# 4. Ambiguous cases (2000)
# - Partial matches, related but not exact
# - Example: Q="When created?" A="1996" Context="FEL created 1996"
# - Label: 0.3-0.6 (medium relevance - correct type, wrong entity)
```

### Example Training Samples

```json
{
  "question": {
    "text": "Kiu kreis Esperanton?",
    "ast": {...},
    "question_type": "WHO"
  },
  "answer": {
    "text": "Zamenhof",
    "ast": {...},
    "entity_type": "PERSON_NAME"
  },
  "context": {
    "text": "Zamenhof kreis Esperanton en 1887",
    "ast": {...}
  },
  "label": 1.0,
  "rationale": "Exact match: Zamenhof created Esperanto"
}

{
  "question": {
    "text": "Kiu kreis Esperanton?",
    "ast": {...}
  },
  "answer": {
    "text": "Esperantista",
    "ast": {...},
    "entity_type": "PERSON_GENERIC"
  },
  "context": {
    "text": "esperantistoj en Berlino",
    "ast": {...}
  },
  "label": 0.05,
  "rationale": "Wrong context: about Esperantists in Berlin, not creation"
}

{
  "question": {
    "text": "Kiam estis kreita Esperanto?",
    "ast": {...}
  },
  "answer": {
    "text": "1996",
    "ast": {...}
  },
  "context": {
    "text": "FEL estis kreita en 1996",
    "ast": {...}
  },
  "label": 0.15,
  "rationale": "Correct type (year), wrong entity (FEL not Esperanto)"
}
```

### 1. Data Preparation
- [ ] Collect positive (Q, A, C) triples from correct extractions
- [ ] Generate hard negatives (wrong answers from same document)
- [ ] Generate easy negatives (random wrong answers)
- [ ] Label ambiguous cases with relevance scores
- [ ] Train/val/test splits

### 2. Model Implementation
- [ ] Implement `QuestionEncoder`
- [ ] Implement `AnswerEncoder`
- [ ] Implement `ContextEncoder`
- [ ] Implement `CrossAttention` layers
- [ ] Implement `VerificationFeatureExtractor`
- [ ] Implement `AnswerVerifier` main model

### 3. Training
- [ ] Binary cross-entropy loss (relevance prediction)
- [ ] Validation monitoring (precision, recall, F1)
- [ ] ROC-AUC tracking
- [ ] Early stopping
- [ ] Model checkpointing

### 4. Integration
- [ ] Add to `ASTAnswerExtractor`
- [ ] Re-rank candidates by verification score
- [ ] Combine with entity type + role scores
- [ ] Make optional (flag: `use_answer_verifier`)

### 5. Evaluation
- [ ] Standalone: precision/recall on verification task
- [ ] Integration: RAG benchmark final accuracy
- [ ] Error analysis: where does it still fail?

## Acceptance Criteria

- [ ] Model precision >85% at 0.5 threshold
- [ ] Recall >80% (catches most wrong answers)
- [ ] AST-based encoding (no raw text reliance)
- [ ] Structured output with explanations
- [ ] Integrated into extraction pipeline
- [ ] RAG accuracy improves by 5-10% over Phase 2
- [ ] Model size <20M parameters

## Expected Impact

**Before:**
- "Esperantista" from wrong context passes all checks
- Accuracy: ~70-75% (after Phase 2)

**After:**
- "Esperantista" gets 0.05 relevance score → rejected
- Correct answer from lower-ranked doc promoted
- Accuracy: ~80-85%

## Time Estimate

**2 weeks**
- Week 1: Data generation, model implementation, initial training
- Week 2: Hard negative mining, tuning, integration, evaluation

**Comment by marctjones** [comment]
Closing as part of a 2026-09-08 backlog cleanup (see PLAN.md, DESIGN.md, CLAUDE.md's merge gate).

This issue predates the merge gate (adopted 2026-07-13) and the AST-native orchestrator rewrite. It describes work against the pre-orchestrator RAG pipeline or the learned model stack (M1/M2 embeddings, compositional embeddings, entity classifiers, neural rerankers) that was pruned from the repo -- klareco/embeddings/, klareco/models/, and klareco/summarization/ no longer exist. It has no milestone and was never benchmarked, matching exactly the failure mode CLAUDE.md documents (25 capability issues filed 2026-05-26, 8 implemented, none benchmarked).

Labeled archived-2026-09 so this batch is recoverable: gh issue list --state closed --label archived-2026-09. If this capability is still wanted, re-file it against the current architecture naming the benchmark metric and test the merge gate requires.
